flutter_install_type.go

```diff
@@ -0,0 +1,279 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/bitrise-io/go-utils/v2/command"
+)
+
+const (
+	FVMName              = "fvm"
+	ASDFName             = "asdf"
+	ManualName           = "manual"
+	FVMCacheVersionsPath = "/fvm/versions"
+	FVMCacheDefaultPath  = "/fvm/default/bin/flutter"
+	ASDFShimsPath        = "/.asdf/shims/flutter"
+)
+
+type FlutterInstallType struct {
+	Name                     string
+	IsAvailable              bool                                          // if the tool is available, this will be set to true later
+	InstalledVersionsCommand func() *command.Command                       // command to list available versions installed by the tool
+	ReleasesCommand          func(version flutterVersion) *command.Command // command to list available releases (if applicable)
+	Install                  func(version flutterVersion) error            // function to install a specific version
+	SetDefault               func(version flutterVersion) error            // function to set a specific version as default (if applicable)
+}
+
+func (f *FlutterInstaller) NewFlutterInstallTypeFVM() FlutterInstallType {
+	cmd := f.CmdFactory.Create("fvm", []string{"--version"}, nil)
+	f.Donef("$ %s", cmd.PrintableCommandArgs())
+	versionOut, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		f.Warnf("fvm version manager is not available")
+		return FlutterInstallType{
+			Name:        FVMName,
+			IsAvailable: false,
+		}
+	} else {
+		f.Debugf("fvm --version output: %s", versionOut)
+	}
+
+	after3_0_0, after3_1_0, after3_2_1, err := fvmParseVersionAndFeatures(versionOut)
+	if err != nil {
+		f.Warnf("Failed to investigate FVM version: %s", err)
+	}
+	listArgs := []string{"list"}
+	if after3_1_0 {
+		listArgs = []string{"api", "list", "--skip-size-calculation"}
+	}
+
+	defaultArgs := []string{}
+	if after3_2_1 {
+		defaultArgs = append(defaultArgs, "--fvm-skip-input")
+	}
+
+	return FlutterInstallType{
+		Name:        FVMName,
+		IsAvailable: true,
+		InstalledVersionsCommand: func() *command.Command {
+			cmd := f.CmdFactory.Create("fvm", listArgs, nil)
+			f.Donef("$ %s", cmd.PrintableCommandArgs())
+			return &cmd
+		},
+		Install: func(version flutterVersion) error {
+			args := append([]string{"install", fvmCreateVersionString(version)}, defaultArgs...)
+			if after3_0_0 {
+				args = append(args, "--setup")
+			}
+
+			cmd := f.CmdFactory.Create("fvm", args, nil)
+			f.Donef("$ %s", cmd.PrintableCommandArgs())
+			if out, err := cmd.RunAndReturnTrimmedCombinedOutput(); err != nil {
+				return fmt.Errorf("install: %s %s", err, out)
+			} else {
+				f.Debugf("Installed Flutter: %s", out)
```

Comment:
Maybe an explanation comment here about what we ran into and why we are updating $PATH